packages/flutter_tools/lib/src/commands/update_packages.dart

```diff
@@ -60,3 +102,418 @@ class UpdatePackagesCommand extends FlutterCommand {
   }
 }
+
+enum Section { dependencies, devDependencies, dependencyOverrides, other }
+
+const String kTransitiveMagicString = '# TRANSITIVE DEPENDENCY';
+
+class PubspecYaml {
+  factory PubspecYaml(Directory directory) {
+    final File file = _pubspecFor(directory);
+    return new PubspecYaml._(file, _parse(file.path, file.readAsLinesSync()));
+  }
+
+  PubspecYaml._(this.file, this.inputData);
+
+  final File file;
+
+  final List<PubspecLine> inputData;
+
+  static List<PubspecLine> _parse(String filename, List<String> lines) {
+    final List<PubspecLine> result = <PubspecLine>[];
+    Section section = Section.other;
+    bool seenMain = false;
+    bool seenDev = false;
+    final Map<String, PubspecDependency> masterDependencies = <String, PubspecDependency>{};
+    PubspecDependency lastDependency;
+    for (String line in lines) {
+      if (lastDependency != null) {
+        if (line.trim().isNotEmpty) {
+          if (!lastDependency.parseLock(line, filename, lockIsOverride: section == Section.dependencyOverrides)) {
+            result.removeLast();
+            result.add(new PubspecLine(lastDependency.line));
+            result.add(new PubspecLine(line));
+          }
+          lastDependency = null;
+        }
+      } else {
+        final PubspecHeader header = PubspecHeader.parse(line);
+        if (header != null) {
+          section = header.section;
+          if (section == Section.devDependencies) {
+            if (seenDev)
+              throw 'Two dev_dependencies sections found in $filename. There should only be one.';
+            seenDev = true;
+          } else if (section == Section.dependencies) {
+            if (seenMain)
+              throw 'Two dependencies sections found in $filename. There should only be one.';
+            if (seenDev)
+              throw 'The dependencies section was after the dev_dependencies section in $filename. To enable one-pass processing, the dependencies section must come before the dev_dependencies section.';
+            seenMain = true;
+          }
+          result.add(header);
+        } else if (section == Section.other) {
+          result.add(new PubspecLine(line));
+        } else {
+          final PubspecDependency dependency = PubspecDependency.parse(line, filename: filename);
+          if (dependency != null) {
+            result.add(dependency);
+            if (dependency.kind == DependencyKind.unknown)
+              lastDependency = dependency;
+            if (section != Section.dependencyOverrides) {
+              masterDependencies[dependency.name] = dependency;
+            } else {
+              masterDependencies[dependency.name]?.markOverridden(dependency);
+            }
+          } else {
+            result.add(new PubspecLine(line));
+          }
+        }
+      }
+    }
+    return result;
+  }
+
+  Set<PubspecDependency> get dependencies {
```

Comment:
This could also use a doc comment as it's not a trivial getter. I think you are flattening and validating dependencies here.

Reply:
added comments